[clang][test] Rewrote test using command substitution to work with lit internal shell syntax

connieyzhu opened the pull request:
This patch rewrites a test that uses command substitution `$()` and the `stat` command, which are not supported by lit's internal shell. Instead of using this syntax to perform the file size comparison done in this test, a Python script is used instead to perform the same operation.

Fixes https://github.com/llvm/llvm-project/issues/102384.

connieyzhu committed:
[clang][test] Rewrote test to work with lit internal shell syntax

This patch rewrites a test that uses command substitution $() and the
stat command, which are not supported by lit's internal shell. Instead
of using this syntax to perform the file size comparison done in this
test, a Python script is used instead to perform the same operation.

ilovepi commented on clang/test/Modules/compare-file-size.py:
```diff
@@ -0,0 +1,22 @@
+        print(f"Unable to get file size of {file_path}")
+        return None
+
+def main():
```
Please add some documentation about what this function does, and its usage. Other python files used in tests should give you an idea about what's appropriate.

ilovepi commented on clang/test/Modules/compare-file-size.py:
```diff
@@ -0,0 +1,22 @@
+import argparse
```
does this require a license/copyright header?  

I see a `#!/usr/bin/env python` in https://github.com/llvm/llvm-project/blob/cdd11d694a406a98a16d6265168ee2fbe1b6a87c/llvm/utils/lit/tests/Inputs/shtest-env/print_environment.py#L1, do yo need something similar here? You're using python to invoke, so it's probably fine, but it seems a bit inconsistent. Do you know what's more typical in the test support scripts like this?

ilovepi replied:
@pogo59 You mentioned in the [RFC](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179/29) that when there isn't another good option that a python script may be an OK substitute in some cases. I'm only aware of a couple of those in the tree(mostly in lit's own test directory). Do you know if these require a license header, or if we have a good convention for wether to use `#!/bin/env python` or not.

pogo59 replied:
There is a lit substitution `%python` and I see a fair number of these in the lit tests. Some use inline python scripts, some have separate scripts, and a number of them have the python script directly in the test file (`RUN: %python %s ...`). It's far more common in llvm than clang, so you may have better luck looking there for inspiration.

Test files generally do not have license headers.

ilovepi replied:
> Test files generally do not have license headers. 

Ah, right. I was thinking of them differently, since they're not par to of the test per se, but now that you spell it out, that makes sense. Thanks.

ilovepi commented on clang/test/Modules/compare-file-size.py:
```diff
@@ -0,0 +1,22 @@
+import os
+
+def get_file_size(file_path):
+    try:
+        return os.path.getsize(file_path)
+    except:
+        print(f"Unable to get file size of {file_path}")
+        return None
```
Is the `try/except` better than just letting the exception propagate up through main? I'm fine w/ it, but I wonder if `print` is better than the unhandled exception? it also doesn't fail, does it? So I don't think you'd return an error code if this didn't succeed, would you?

connieyzhu replied:
Now that I look at it again, I think having an unhandled exception would be better. I didn't realize that my implementation doesn't actually return an error code, but I would think having the error code would be better for debugging.

ChuanqiXu9 reviewed (approved):
LGTM on my side. Thanks. But I didn't take a detailed look into https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179/29. So it might be good to wait for people there to approve this

connieyzhu committed:
[clang][test] Modified python program to return error code when failing
to get file size

This patch removes the print statement that executes in the case of
exceptions, opting to let the system return its own error code when
failing to get the file size of a certain file. There are also some NFC
changes: adding description for compare-file-size.py and changing the
%{python} syntax to use the exisitng lit substitution %python.

connieyzhu committed:
[clang][test][NFC] Edited comment to have clearer description

connieyzhu committed:
[clang][test] Fixed faulty Python file syntax

connieyzhu committed:
[NFC] Python styling fixes
